rfc/rfc-56/rfc-56.md

```diff
@@ -0,0 +1,238 @@
+
+This design expands the create marker API on timeline server.
+
+![](figure3.png)
+
+As shown in Figure 3, the client side calls create marker API, requesting to create markers.
+
+On the timeline server side, there is a MarkerCheckerRunnable thread, which checks the conflicts between the markers in
+memory created by the current writer and the markers persisted to the storage created by other active writers
+periodically. If any conflict is detected, this thread updates the value of hasConflict, which is used to respond to the
+marker creation requests.
+
+We did a performance test for async conflict checker using local FS and two concurrency writers based on 1500+ inflight
+markers located in MARKER0~MARKER19. It takes 179ms to finish checking conflict.
+
+Here we need to pay attention to the following details:
+
+Firstly, There may be a concurrency issue between reading and writing the same MARKER file causing EOFException. In
+order to simplify the model, we return empty result of markers when catching any exception during reading specific
+MARKER and let the subsequent checker to make up for it. For example, writer1 reads writer2's MARKER0 file, but at this
+moment writer2 is overwriting MARKER0 file. Finally, we get empty result after writer1 reading writer2's MARKER0. In
+extreme cases, the delay of early conflict detection will also occur. Fortunately, we can adjust the frequency of the
+schedule according to the actual situation.
+
+Secondly, the marker checker runnable only looks at the instants smaller than the current writer for conflict detection.
+This is to avoid two writers discovering conflicts with each other at the same time and failing together. Such conflict
+detection logic based on the instant time is consistent with the existing pre-commit conflict detection.
+
+### [2] Check Commit Conflict: Why we still need to check commit conflict here?
+
+As we know, when a writer completes writing and commits, the corresponding markers are deleted immediately. Other
+writers miss the conflict detection if solely based on the markers.
+
+Let's take Figure 4 as an example
+
+![](figure4.png)
+
+Writer1 starts writing data at time t1 and finishes at time t3. Writer2 starts writing at time t2, and updates a file
+already updated by writer1 at time t4. Since all markers of writer1 have been deleted at time t4, such conflict cannot
+be found in the stage of marker conflict detection until starting to commit. In order to avoid such delay of early
+conflict detection, it is necessary to add the steps of checking commit conflict during the detection.
+
+```
+1. Get and cache all instants before init write handler as set1.
+2. Create and reload a new active timeline after finishing marker conflict checking as set2.
+3. Get the diff instants between set1 and set2 as set3.
+4. Read all the changed fileIds contained in set3 as fileId set1
+5. Check if current writing fileID set2 is overlapped with fileId set1
```

Comment:
also since OCC Early detection is pure optimization play - Can we skip on these complex logic and keep the design simple. Worst case there are some false positives that could go through till pre-commit check. 

Personally, I would rather have OCC early detection logic very simple and catch 95% of the cases than being very complicated and catch 99% of the cases.

Reply:
Maybe we can set a functional flag default as false.
For common users they don't need to care about this flag.
For users who are more sensitive to this early conflict detect, they can enable this hard check.
Hi @yihua what do you think here  :)

Reply:
Thats sounds okay to me.

Reply:
Without checking the metadata of completed commits from other concurrent writers, we miss the type of conflicts happening between t3 and t10 in the case below, which I think happens much more than 5% of the cases if there are conflicts, and lead to resource waste.  For timeline-server-based markers, since the timeline server is going to do such detection asynchronously, it should not add latency to the hot path of writer data files.
```
writer 1: t1-----t3
writer 2:     t2---------------<conflict>------------------t10
```

IMO, given the commit metadata check is sth pre-commit conflict detection also does, we can reuse the same logic and generalize it for the early conflict detection as well.  I agree that we can add a feature flag for such logic, but we can keep the default as true for timeline-server-based markers and false for direct markers.  If the user really sees performance issues, they can turn it off.  What do you guys think?

Reply:
LGTM !